Chapter 4 Task 20

tasiq opened the pull request.

tasiq committed:
Chapter 4 Task 20

EvgeneZhurov commented on src/main/java/com/walking/intensive/chapter4/task20/Main.java:
```diff
@@ -1,7 +1,84 @@
+        doHeapSort(array, root, rightBorder - 1);
+    }
+
+    public static void heapify(int[] array, int root, int rightBorder) {
```
heapify - старайтесь давать методам максимально информативные имена.

EvgeneZhurov commented on src/main/java/com/walking/intensive/chapter4/task20/Main.java:
```diff
@@ -1,7 +1,84 @@
+
+        int i = 0;
+
+        while (isSorted && i < array.length - 1) {
```
между циклами for и while есть разница. Вы знаете точное количество итераций цикла, в нем тоже есть логический блок выхода из цикла и логические условия можно перечислить в нем. Лучше for в данном случае.

EvgeneZhurov commented on src/main/java/com/walking/intensive/chapter4/task20/Main.java:
```diff
@@ -1,7 +1,84 @@
     public static void main(String[] args) {
 
+        int[] array1 = {1, 2, 3, 4, 5};
+        int[] array2 = {1, 5, 7, 6, 45, 4, 31, 9, 8, 3, 2};
```
int[] array2 = {1, 5, 7, 6, 45, 4, 31, 9, 8, 3};
метод падает

tasiq committed:
Chapter 4 Task 20:

- метод hepify переименован на restoreHeap;
- в методе heapSort проверка на отсортированность переданного массива заменена с перебора циклом while на цикл for;
- исправлена ошибка в методе buildHeap, теперь в нем метод restoreHeap вызывается с корректным значением правой границы массива array.length-1.

nb: непонятно, почему оно вообще работало раньше, если постоянно пыталось обратиться к элементу массива array.length

EvgeneZhurov commented on src/main/java/com/walking/intensive/chapter4/task20/Main.java:
```diff
@@ -1,7 +1,101 @@
+        //int i = 0;
+
+        for (int i = 0; i < array.length - 2 && isSorted; i++) {
+            if (array[i] > array[i + 1]) {
```
идея советует ставить break, чтобы не мучиться с циклом

EvgeneZhurov reviewed (approved):
респект
